crates/league-modpkg/src/read.rs

```diff
@@ -21,64 +19,77 @@ impl Modpkg {
             return Err(ModpkgError::InvalidVersion(version));
         }
 
-        let name = reader.read_len_prefixed_string::<LE>()?;
-        let display_name = reader.read_len_prefixed_string::<LE>()?;
-        let description = reader.read_len_prefixed_string::<LE>()?;
-        let version = reader.read_len_prefixed_string::<LE>()?;
-        let distributor = reader.read_len_prefixed_string::<LE>()?;
-
-        let authors = Self::read_authors(reader)?;
-        let license = ModpkgLicense::read(reader)?;
-        let chunks = Self::read_chunks(reader)?;
-        Ok(Self {
-            name,
-            display_name,
-            description: match description.len() {
-                0 => None,
-                _ => Some(description),
-            },
-            version,
-            distributor: match distributor.len() {
-                0 => None,
-                _ => Some(distributor),
-            },
-            authors,
-            license,
-            chunks,
-        })
-    }
+        let header_size = reader.read_u32::<LE>()?;
+        let metadata_size = reader.read_u32::<LE>()? as usize;
+        let signature_size = reader.read_u32::<LE>()? as usize;
+        let chunk_paths_size = reader.read_u32::<LE>()? as usize;
+        let wad_paths_size = reader.read_u32::<LE>()? as usize;
+        let layers_size = reader.read_u32::<LE>()? as usize;
+        let chunk_count = reader.read_u32::<LE>()?;
 
-    fn read_authors(reader: &mut BufReader<impl Read>) -> Result<Vec<ModpkgAuthor>, ModpkgError> {
-        let count = reader.read_u32::<LE>()?;
-        let mut authors = Vec::with_capacity(count as usize);
-        for _ in 0..count {
-            let name = reader.read_len_prefixed_string::<LE>()?;
-            let role = reader.read_len_prefixed_string::<LE>()?;
-
-            authors.push(ModpkgAuthor {
-                name,
-                role: match role.len() {
-                    0 => None,
-                    _ => Some(role),
-                },
+        let current_pos = reader.seek(SeekFrom::Current(0))?;
```

Comment:
I wouldn't assume the reader starts at 0. You could use the measure lib function but you'd have to pack/unpack all the fields as a tuple or mutate or something, so might be easier to just measure the cursor pos before reading anything.